python/ql/test/experimental/dataflow/coverage/argumentPassing.py

```diff
@@ -183,7 +165,86 @@ def with_kw(a=""):
 
 
 def test_kw_doublestar():
-    def with_doublestar(**a):
-        SINK1(a["a"])
+    def with_doublestar(**kwargs):
+        SINK1(kwargs["a"])
 
     with_doublestar(a=arg1)  #$ arg1 func=test_kw_doublestar.with_doublestar
+
+
+def only_kwargs(**kwargs):
+    SINK1(kwargs["a"])
+    SINK2(kwargs["b"])
+    SINK3_F(kwargs["c"])
+
+@expects(3)
+def test_kwargs():
+    args = {"a": arg1, "b": arg2, "c": "safe"} # $ arg1 arg2 func=only_kwargs
+    only_kwargs(**args)
+
+
+def mixed(a, **kwargs):
+    SINK1(a)
+    try:
+        SINK1_F(kwargs["a"]) # since 'a' is a keyword argument, it cannot be part of **kwargs
+    except KeyError:
+        print("OK")
+    SINK2(kwargs["b"])
+    SINK3_F(kwargs["c"])
+
+@expects(4*3)
+def test_mixed():
+    mixed(a=arg1, b=arg2, c="safe") # $ arg1 arg2
+
+    args = {"b": arg2, "c": "safe"} # $ arg2 func=mixed
+    mixed(a=arg1, **args) # $ arg1
+
+    args = {"a": arg1, "b": arg2, "c": "safe"} # $ arg1 arg2 func=mixed
+    mixed(**args)
+
+
+def starargs_only(*args):
+    SINK1(args[0])
+    SINK2(args[1])
+    SINK3_F(args[2])
+
+@expects(3*3)
+def test_only_starargs():
+    starargs_only(arg1, arg2, "safe") # $ arg1 arg2 SPURIOUS: bad2,bad3="arg1" bad1,bad3="arg2"
+
+    args = (arg2, "safe") # $ arg2 func=starargs_only SPURIOUS: bad1,bad3="arg2"
+    starargs_only(arg1, *args) # $ arg1 SPURIOUS: bad2,bad3="arg1"
+
+    args = (arg1, arg2, "safe") # $ arg1 arg2 func=starargs_only
+    starargs_only(*args)
+
```

Comment:
I've added that test in d9fbe58

It doesn't quite work that way, since we only treat the first `*args` as an argument -- see https://github.com/github/codeql/blob/d9fbe58ad5a4f515806de4c4baeeb43cde336bc5/python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll#L1088-L1096

Reply:
I see, so this is better than I feared, but the test is still interesting.